src/algebra/discrete-log.md

```diff
@@ -113,32 +109,96 @@ This can be done by keeping a variable that is multiplied by $a$ each time we in
 With this change, the complexity of the algorithm is still the same, but now the $\log$ factor is only for the `map`.
 Instead of a `map`, we can also use a hash table (`unordered_map` in C++) which has the average time complexity $O(1)$ for inserting and searching.
 
-```cpp
+Problems often ask for the minimum $x$ which satisfies the solution.  
+It is possible to get all answers and take the minimum, or reduce the first found answer using [Euler's theorem](./algebra/phi-function.html#toc-tgt-2), but we can be smart about the order in which we calculate values and ensure the first answer we find is the minimum.
+
+```cpp discrete_log
+// Returns minimum x for which a ^ x % m = b % m, a and m are coprime.
 int solve(int a, int b, int m) {
+	a %= m, b %= m;
 	int n = (int) sqrt (m + .0) + 1;
 
 	int an = 1;
 	for (int i = 0; i < n; ++i)
 		an = (an * 1ll * a) % m;
 
-	map<int, int> vals;
-	for (int p = 1, cur = an; p <= n; ++p) {
-		if (!vals.count(cur))
-			vals[cur] = p;
+	unordered_map<int, int> vals;
+	for (int q = 0, cur = b; q <= n; ++q) {
+		vals[cur] = q;
+		cur = (cur * 1ll * a) % m;
+	}
+
+	for (int p = 1, cur = 1; p <= n; ++p) {
 		cur = (cur * 1ll * an) % m;
+		if (vals.count(cur)) {
+			int ans = n * p - vals[cur];
+			return ans;
+		}
 	}
+	return -1;
+}
+```
+
+The complexity is $O(\sqrt{m})$ using `unordered_map`.
+
+## When $a$ and $m$ are not coprime
+Let $g = \gcd(a, m)$, and $g > 1$. Clearly $a^x \bmod m$ for every $x \ge 1$ will be divisible by $g$.
+
+If $g \nmid b$, there is no solution for $x$.
 
+If $g \mid b$, let $a = g \alpha, b = g \beta, m = g \nu$.
+$$
+\begin{aligned}
+a^x & \equiv b \mod m \\\
+(g \alpha) a^{x - 1} & \equiv g \beta \mod g \nu \\\
+\alpha a^{x-1} & \equiv \beta \mod \nu
+\end{aligned}
+$$
+
+The baby step giant step algorithm can be easily extended to solve $ka^{x} \equiv b \mod m$ for $x$.
+```cpp discrete_log_extended
+// Returns minimum x for which a ^ x % m = b % m.
+int solve(int a, int b, int m) {
+	a %= m, b %= m;
+	int k = 1, add = 0, g;
+	while ((g = gcd(a, m)) > 1) {
+		if (b == k)
+			return add;
+		if (b % g)
+			return -1;
+		b /= g, m /= g, ++add;
+		k = (k * 1ll * a / g) % m;
+	}
+
+	int n = (int) sqrt (m + .0) + 1;
```

Comment:
This'll only convert `m` to double isn't it? I don't think this is required as since `sqrt(m) + 1` or `sqrt(m) + 2` should suffice anyway. If we want to have a check like this, do you think it'd be better to do `sqrt(m + .5)` instead to avoid precision issue when m is close to a large perfect square?

Reply:
Agree this is redundant, according to [cppreference](https://en.cppreference.com/w/cpp/numeric/math/sqrt) using an integral type is equivalent to casting it to double. Updated to `sqrt(m) + 1`. 

Do you suggest `sqrt(m + .5)` because it gives the exact square root? The algorithm works even if n is 1 more (or any amount more in fact) than sqrt(n), so I think it should be fine.

Reply:
I'm fine either ways, doesn't look like it's going to be an issue so it can stay like this.